### ISSUE TITLE
Questions in merge part

### ISSUE THREAD

**Issue by argosdh** [liuzhengzhe/One-Thing-One-Click#14]
Hi, I have several questions in merge section. From my understanding, the merge part does the graph propagation based on results of 3D-Unet and RelationNet. However, in ./merge/model/pointgroup/pointgroup.py, only U and U_prod were used to generate avg and conf. Other variables like result_avg_crf and result_avg were never used. And I'm also confused about the algorithm of graph propagation. In  ./merge/test_train.py, Q[i] means the ith super voxel's class semantic score calculated by production of super voxel pooled 3D-Unet semantic prediction U(M, 20) and relation net prediction of each labelled super voxel U_prod(M, 20). In U_prod, only those super voxels with labels are scored, other lines are all zeros. Why you use softmax(U*U_prod*5) as the final class prediction of each super voxel, I can't connect this line with the paper. Also, the confidence(conf) is calculated by torch.max(torch.cat(U,U_prod)). Are they comparable? Just choose the biggest score out of 40 scores and why this largest score is the one which means to be the confidence of super voxel SV belongs to class C?

**Comment by liuzhengzhe** [comment]
Thanks for your message. 
1. Other variables like result_avg_crf and result_avg: I used them to check intermediate results and see how these operations influence the performance. 
2. softmax(U * U_prod * 5) as the final class prediction: We have mentioned it in paper page 5 "The prediction of relation network is further combined with the prediction of 3D U-Net by multiplying the predicted possibilities of each category to boost the performance" 
3. the confidence(conf) is calculated by torch.max(torch.cat(U,U_prod)): we use the simple thresholding way to update pseudo label, and the maximum value is the prediction confidence.

**Comment by derkaczda** [comment]
About 2.: Why multiply by 5, this seems like a random number to me

**Comment by liuzhengzhe** [comment]
It is a hyper parameter, and can be set to other values.

derkaczda ***@***.***> 于2022年2月10日周四 00:13写道：

**Comment by c6376315qqso** [comment]
> Thanks for your message.
> 
> 1. Other variables like result_avg_crf and result_avg: I used them to check intermediate results and see how these operations influence the performance.
> 2. softmax(U * U_prod * 5) as the final class prediction: We have mentioned it in paper page 5 "The prediction of relation network is further combined with the prediction of 3D U-Net by multiplying the predicted possibilities of each category to boost the performance"
> 3. the confidence(conf) is calculated by torch.max(torch.cat(U,U_prod)): we use the simple thresholding way to update pseudo label, and the maximum value is the prediction confidence.

Hi, I can't see where the crf is in the code, it seems like in the code you didn't use a crf to predict labels?

**Comment by liuzhengzhe** [comment]
We implement CRF using  pytorch. Please check
merge/model/pointgroup/pointgroup.py: line 383 to 402.
